search/mongo_search.go

```diff
@@ -66,11 +76,25 @@ func (m *MongoSearcher) Search(query Query) (results interface{}, total uint32,
 	// build the BSON query (without any options)
 	bsonQuery := m.convertToBSON(query)
 
+	// Check to see if we already have a count cached for this query. If so, use it
+	// and tell the searcher to skip doing the count. This can only be done reliably if
+	// the server is in -readonly mode.
+	docount := true
+	queryHash := fmt.Sprintf("%x", md5.Sum([]byte(query.Query)))
```

Comment:
I would declare `queryHash` here, but wait to calculate the hash until you are in the readonly block. That way, you aren't calculating hashes you will never use in read/write mode.